src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -0,0 +1,945 @@
+        protected readonly SyntaxReference syntaxReferenceOpt;
+        protected SourceMethodSymbolWithAttributes(SyntaxReference syntaxReferenceOpt)
+        {
+            this.syntaxReferenceOpt = syntaxReferenceOpt;
+        }
+
+#nullable enable
+        internal CSharpSyntaxNode? GetBinderContextNode()
+        {
+            switch (SyntaxNode)
+            {
+                case ConstructorDeclarationSyntax constructor:
+                    return constructor.Initializer ?? (CSharpSyntaxNode?)constructor.Body ?? constructor.ExpressionBody;
+                case BaseMethodDeclarationSyntax method:
+                    return (CSharpSyntaxNode?)method.Body ?? method.ExpressionBody;
+                case AccessorDeclarationSyntax accessor:
+                    return (CSharpSyntaxNode?)accessor.Body ?? accessor.ExpressionBody;
+                case ArrowExpressionClauseSyntax arrowExpression:
+                    Debug.Assert(arrowExpression.Parent.Kind() == SyntaxKind.PropertyDeclaration ||
+                                 arrowExpression.Parent.Kind() == SyntaxKind.IndexerDeclaration);
+                    return arrowExpression;
+                case LocalFunctionStatementSyntax localFunction:
+                    return (CSharpSyntaxNode?)localFunction.Body ?? localFunction.ExpressionBody;
+                default:
+                    return null;
+            }
+        }
+#nullable restore
+
+        internal SyntaxReference SyntaxRef
+        {
+            get
+            {
+                return this.syntaxReferenceOpt;
+            }
+        }
+
+        internal CSharpSyntaxNode SyntaxNode
+        {
+            get
+            {
+                return (this.syntaxReferenceOpt == null) ? null : (CSharpSyntaxNode)this.syntaxReferenceOpt.GetSyntax();
+            }
+        }
+
+        internal SyntaxTree SyntaxTree
+        {
+            get
+            {
+                return this.syntaxReferenceOpt == null ? null : this.syntaxReferenceOpt.SyntaxTree;
+            }
+        }
+
+        public override FlowAnalysisAnnotations ReturnTypeFlowAnalysisAnnotations =>
+            DecodeReturnTypeAnnotationAttributes(GetDecodedReturnTypeWellKnownAttributeData());
+
+        public override ImmutableHashSet<string> ReturnNotNullIfParameterNotNull
+            => GetDecodedReturnTypeWellKnownAttributeData()?.NotNullIfParameterNotNull ?? ImmutableHashSet<string>.Empty;
+
+        /// <summary>
+        /// Symbol to copy bound attributes from, or null if the attributes are not shared among multiple source method symbols.
+        /// </summary>
+        /// <remarks>
+        /// Used for example for event accessors. The "remove" method delegates attribute binding to the "add" method.
+        /// The bound attribute data are then applied to both accessors.
+        /// </remarks>
+        protected virtual SourceMethodSymbolWithAttributes BoundAttributesSource
```

Comment:
>SourceMethodSymbolWithAttributes [](start = 26, length = 32)

Is there a need to change this type?

Reply:
There is not. Should I revert the return type to `SourceMemberMethodSymbol`?

---
In reply to: [354029793](https://github.com/dotnet/roslyn/pull/40011#discussion_r354029793) [](ancestors = 354029793)

Reply:
> There is not. Should I revert the return type to SourceMemberMethodSymbol?

I would avoid making any changes without a benefit.

---
In reply to: [355080054](https://github.com/dotnet/roslyn/pull/40011#discussion_r355080054) [](ancestors = 355080054,354029793)